# CARICAMENTO DF GREZZI DI ISTAT E SITUAS

In [1]:
import pandas as pd
import os

In [2]:
os.getcwd()

'c:\\Users\\eliav\\Desktop\\Road_accidents_analysis\\Road_accidents_analysis_FIXED'

In [3]:
df_situas_raw = pd.read_csv("Df_raw/df_situas_raw.csv")
df_situas_raw

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune (alfanumerico),Codice Comune (numerico),Comune,Comune (dizione straniera),Sigla automobilistica,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente),TIME_PERIOD,Codice Provincia (Storico)
0,1,1,1,1001,1001,Agliè,NaN,TO,0,0,2574.0,2001,"13,1462",2001,2557,2001,2001,NaN
1,1,1,1,1002,1002,Airasca,NaN,TO,0,0,3554.0,2001,"15,7393",2001,3543,2001,2001,NaN
2,1,1,1,1003,1003,Ala di Stura,NaN,TO,0,0,479.0,2001,"46,3315",2001,480,2001,2001,NaN
3,1,1,1,1004,1004,Albiano d'Ivrea,NaN,TO,0,0,1696.0,2001,"11,7315",2001,1687,2001,2001,NaN
4,1,1,1,1005,1005,Alice Superiore,NaN,TO,0,0,616.0,2001,"7,3796",2001,619,2001,2001,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
192726,5,20,111,111103,111103,Villaputzu,NaN,SU,0,0,4509.0,2021,"181,3947",2024,4368,2024,2024,111.0
192727,5,20,111,111104,111104,Villasalto,NaN,SU,0,0,989.0,2021,"130,3596",2024,903,2024,2024,111.0
192728,5,20,111,111105,111105,Villasimius,NaN,SU,0,0,3705.0,2021,"58,1781",2024,3735,2024,2024,111.0
192729,5,20,111,111106,111106,Villasor,NaN,SU,0,0,6599.0,2021,"86,8137",2024,6534,2024,2024,111.0


In [4]:
#controllo per vedere se per un comune sono associati 24 anni
df_situas_raw[df_situas_raw["Codice Comune (numerico)"] == 1001][["Codice Comune (numerico)", "TIME_PERIOD"]]

,Codice Comune (numerico),TIME_PERIOD
0,1001,2001
8102,1001,2002
16204,1001,2003
24304,1001,2004
32405,1001,2005
40506,1001,2006
48607,1001,2007
56708,1001,2008
64809,1001,2009
72909,1001,2010


In [45]:
df_situas_raw.groupby("Codice Comune (numerico)")["TIME_PERIOD"].nunique().value_counts()

TIME_PERIOD
24    7488
16     145
9      144
8      126
10     123
15     118
5       80
18      65
14      64
13      57
17      43
6       25
11      25
19      23
7       19
23       8
22       7
12       5
2        4
1        4
4        3
20       2
Name: count, dtype: int64

In [6]:
df_istat_raw = pd.read_csv("Df_raw/df_istat_raw.csv")
df_istat_raw

,DATAFLOW,FREQ,REF_AREA,DATA_TYPE,RESULT,TIME_PERIOD,OBS_VALUE,OBS_STATUS,NOTE_DS,NOTE_REF_AREA,NOTE_DATA_TYPE,NOTE_RESULT,NOTE_TIME_PERIOD,BASE_PER,UNIT_MEAS,UNIT_MULT
0,IT1:41_983(1.0),A,1001,KILLINJ,F,2001,10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,IT1:41_983(1.0),A,1001,KILLINJ,F,2002,10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,IT1:41_983(1.0),A,1001,KILLINJ,F,2003,7,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,IT1:41_983(1.0),A,1001,KILLINJ,F,2004,13,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,IT1:41_983(1.0),A,1001,KILLINJ,F,2005,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
573547,IT1:41_983(1.0),A,111107,ROADACC,9,2020,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
573548,IT1:41_983(1.0),A,111107,ROADACC,9,2021,5,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
573549,IT1:41_983(1.0),A,111107,ROADACC,9,2022,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
573550,IT1:41_983(1.0),A,111107,ROADACC,9,2023,4,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [7]:
df_situas_raw.duplicated().sum()

np.int64(0)

In [8]:
df_istat_raw.duplicated().sum()

np.int64(0)

In [9]:
print(df_istat_raw["REF_AREA"].dtype)
print(df_situas_raw["Codice Comune (numerico)"].dtype)
print(df_istat_raw["TIME_PERIOD"].dtype)
print(df_situas_raw["TIME_PERIOD"].dtype)

int64
int64
int64
int64


In [10]:
df_final_raw= pd.merge(df_istat_raw, df_situas_raw, left_on=["REF_AREA","TIME_PERIOD"], right_on=["Codice Comune (numerico)","TIME_PERIOD"], how="left", indicator=True)

df_final_raw

,DATAFLOW,FREQ,REF_AREA,DATA_TYPE,RESULT,TIME_PERIOD,OBS_VALUE,OBS_STATUS,NOTE_DS,NOTE_REF_AREA,...,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente),Codice Provincia (Storico),_merge
0,IT1:41_983(1.0),A,1001,KILLINJ,F,2001,10,NaN,NaN,NaN,...,0.0,0.0,2574.0,2001.0,"13,1462",2001.0,2557.0,2001.0,NaN,both
1,IT1:41_983(1.0),A,1001,KILLINJ,F,2002,10,NaN,NaN,NaN,...,0.0,0.0,2574.0,2001.0,"13,1462",2002.0,2538.0,2002.0,NaN,both
2,IT1:41_983(1.0),A,1001,KILLINJ,F,2003,7,NaN,NaN,NaN,...,0.0,0.0,2574.0,2001.0,"13,1462",2003.0,2588.0,2003.0,NaN,both
3,IT1:41_983(1.0),A,1001,KILLINJ,F,2004,13,NaN,NaN,NaN,...,0.0,0.0,2574.0,2001.0,"13,1462",2004.0,2679.0,2004.0,NaN,both
4,IT1:41_983(1.0),A,1001,KILLINJ,F,2005,2,NaN,NaN,NaN,...,0.0,0.0,2574.0,2001.0,"13,1462",2005.0,2674.0,2005.0,NaN,both
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
573547,IT1:41_983(1.0),A,111107,ROADACC,9,2020,2,NaN,NaN,NaN,...,0.0,0.0,2407.0,2011.0,"27,1937",2020.0,2549.0,2020.0,111.0,both
573548,IT1:41_983(1.0),A,111107,ROADACC,9,2021,5,NaN,NaN,NaN,...,0.0,0.0,2536.0,2021.0,"27,1943",2021.0,2536.0,2021.0,111.0,both
573549,IT1:41_983(1.0),A,111107,ROADACC,9,2022,1,NaN,NaN,NaN,...,0.0,0.0,2536.0,2021.0,"27,1943",2022.0,2575.0,2022.0,111.0,both
573550,IT1:41_983(1.0),A,111107,ROADACC,9,2023,4,NaN,NaN,NaN,...,0.0,0.0,2536.0,2021.0,"27,1943",2023.0,2616.0,2023.0,111.0,both


In [11]:
#dal momento in cui ci sono molte colonne (infatti non le mette tutte "...") guardo 
# se ci sono quelle che mi interessano tra la lista delle colonne
print(df_final_raw.columns.tolist())


['DATAFLOW', 'FREQ', 'REF_AREA', 'DATA_TYPE', 'RESULT', 'TIME_PERIOD', 'OBS_VALUE', 'OBS_STATUS', 'NOTE_DS', 'NOTE_REF_AREA', 'NOTE_DATA_TYPE', 'NOTE_RESULT', 'NOTE_TIME_PERIOD', 'BASE_PER', 'UNIT_MEAS', 'UNIT_MULT', 'Codice Ripartizione geografica', 'Codice Regione', 'Codice Provincia/Uts', 'Codice Comune (alfanumerico)', 'Codice Comune (numerico)', 'Comune', 'Comune (dizione straniera)', 'Sigla automobilistica', 'Capoluogo di Provincia/Uts', 'Capoluogo di Regione', 'Popolazione legale', 'Anno Censimento', 'Superficie (Kmq)', 'Anno (Superficie)', 'Popolazione residente', 'Anno (Popolazione residente)', 'Codice Provincia (Storico)', '_merge']


In [12]:
#Guardo se qualcosa non è stato matchato (quindi NON BOTH)
df_final_raw["_merge"].value_counts()

_merge
both          572700
left_only        852
right_only         0
Name: count, dtype: int64

In [13]:
left_only=df_final_raw[df_final_raw["_merge"]=="left_only"]

display(left_only)

display(f"Ci sono {left_only.shape[0]} rows che non hanno trovato corrispondenza nel df situas, \n sono comuni che hanno subito variazioni amministrative?")

,DATAFLOW,FREQ,REF_AREA,DATA_TYPE,RESULT,TIME_PERIOD,OBS_VALUE,OBS_STATUS,NOTE_DS,NOTE_REF_AREA,...,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente),Codice Provincia (Storico),_merge
65103,IT1:41_983(1.0),A,6064,KILLINJ,F,2019,0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
65122,IT1:41_983(1.0),A,6064,KILLINJ,M,2019,0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
65141,IT1:41_983(1.0),A,6064,ROADACC,9,2019,0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
66828,IT1:41_983(1.0),A,6089,KILLINJ,F,2019,0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
66847,IT1:41_983(1.0),A,6089,KILLINJ,M,2019,0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
566295,IT1:41_983(1.0),A,107022,KILLINJ,M,2016,0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
566306,IT1:41_983(1.0),A,107022,ROADACC,9,2016,3,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
566317,IT1:41_983(1.0),A,107023,KILLINJ,F,2016,0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
566328,IT1:41_983(1.0),A,107023,KILLINJ,M,2016,0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only


'Ci sono 852 rows che non hanno trovato corrispondenza nel df situas, \n sono comuni che hanno subito variazioni amministrative?'

In [14]:
#prendo codice di un comune a caso e vedo qual è il primo anno di NON corrispondenza
codice = left_only["REF_AREA"].iloc[0]
anno = left_only["TIME_PERIOD"].iloc[0]

print(codice, anno)

6064 2019


In [15]:
#guardo se esisteva in df situas --->non esisteva
df_situas_raw[(df_situas_raw["Codice Comune (numerico)"] == codice) &
                (df_situas_raw["TIME_PERIOD"] == anno)]

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune (alfanumerico),Codice Comune (numerico),Comune,Comune (dizione straniera),Sigla automobilistica,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente),TIME_PERIOD,Codice Provincia (Storico)


In [16]:
#guardo quel codice preso come esempio fino a che anno è stato registrato
df_situas_raw[df_situas_raw["Codice Comune (numerico)"] == codice][["Codice Comune (numerico)", "Comune", "TIME_PERIOD"]].sort_values("TIME_PERIOD")

,Codice Comune (numerico),Comune,TIME_PERIOD
920,6064,Cuccaro Monferrato,2001
9022,6064,Cuccaro Monferrato,2002
17124,6064,Cuccaro Monferrato,2003
25224,6064,Cuccaro Monferrato,2004
33325,6064,Cuccaro Monferrato,2005
41426,6064,Cuccaro Monferrato,2006
49527,6064,Cuccaro Monferrato,2007
57628,6064,Cuccaro Monferrato,2008
65729,6064,Cuccaro Monferrato,2009
73829,6064,Cuccaro Monferrato,2010


In [17]:
#creo due set di codici univoci dei due df in maniera tale da sottrarli e vedere se istat ha + o - codici
# rispetto a situas
codici_istat = set(df_istat_raw["REF_AREA"].unique())
codici_situas = set(df_situas_raw["Codice Comune (numerico)"].unique())

print(len(codici_istat - codici_situas))

0


In [18]:
df_situas_raw[
    df_situas_raw["Comune"].str.contains(
        "LU E CUCCARO",
        case=False,
        na=False
    )
][["Comune",
   "Codice Comune (numerico)",
   "TIME_PERIOD"]].sort_values("TIME_PERIOD")

,Comune,Codice Comune (numerico),TIME_PERIOD
146342,Lu e Cuccaro Monferrato,6193,2019
154256,Lu e Cuccaro Monferrato,6193,2020
162159,Lu e Cuccaro Monferrato,6193,2021
170063,Lu e Cuccaro Monferrato,6193,2022
177966,Lu e Cuccaro Monferrato,6193,2023
185866,Lu e Cuccaro Monferrato,6193,2024


In [19]:

primo_left = (
    left_only
    .groupby("REF_AREA")["TIME_PERIOD"]
    .min()
    .reset_index()
    .rename(columns={
        "REF_AREA": "Codice Comune (numerico)",
        "TIME_PERIOD": "Primo left_only"
    })
)

ultimi_anni = (
    df_situas_raw[
        df_situas_raw["Codice Comune (numerico)"].isin(left_only["REF_AREA"])
    ]
    .groupby("Codice Comune (numerico)")["TIME_PERIOD"]
    .max()
    .reset_index()
)

controllo = ultimi_anni.merge(
    primo_left,
    on="Codice Comune (numerico)"
)

controllo["Differenza"] = (
    controllo["Primo left_only"] -
    controllo["TIME_PERIOD"]
)

controllo.sort_values("Differenza")
controllo["Differenza"].value_counts()

Differenza
 1     283
-21      1
Name: count, dtype: int64

La merge è stata effettuata sulla chiave composta (REF_AREA, TIME_PERIOD) e validata tramite validate="many_to_one". Le osservazioni senza corrispondenza (left_only) sono risultate riconducibili quasi esclusivamente a variazioni amministrative dei comuni. Per 283 dei 284 codici comunali coinvolti, il primo anno senza corrispondenza coincide con l'anno successivo all'ultima presenza nel dataset SITUAS, indicando fusioni, soppressioni o cambi di codice. È stato individuato un solo caso anomalo, oggetto di verifica puntuale.

# CREAZIONE DF FINALE CON SOLO COLONNE DI INTERESSE

In [20]:
df_final_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 573552 entries, 0 to 573551
Data columns (total 34 columns):
 #   Column                          Non-Null Count   Dtype   
---  ------                          --------------   -----   
 0   DATAFLOW                        573552 non-null  str     
 1   FREQ                            573552 non-null  str     
 2   REF_AREA                        573552 non-null  int64   
 3   DATA_TYPE                       573552 non-null  str     
 4   RESULT                          573552 non-null  str     
 5   TIME_PERIOD                     573552 non-null  int64   
 6   OBS_VALUE                       573552 non-null  int64   
 7   OBS_STATUS                      0 non-null       float64 
 8   NOTE_DS                         0 non-null       float64 
 9   NOTE_REF_AREA                   0 non-null       float64 
 10  NOTE_DATA_TYPE                  0 non-null       float64 
 11  NOTE_RESULT                     0 non-null       float64 
 12  NOTE_TIME_PER

In [21]:
df_final_raw.isna().sum()

DATAFLOW                               0
FREQ                                   0
REF_AREA                               0
DATA_TYPE                              0
RESULT                                 0
TIME_PERIOD                            0
OBS_VALUE                              0
OBS_STATUS                        573552
NOTE_DS                           573552
NOTE_REF_AREA                     573552
NOTE_DATA_TYPE                    573552
NOTE_RESULT                       573552
NOTE_TIME_PERIOD                  573552
BASE_PER                          573552
UNIT_MEAS                         573552
UNIT_MULT                         573552
Codice Ripartizione geografica       852
Codice Regione                       852
Codice Provincia/Uts                 852
Codice Comune (alfanumerico)         852
Codice Comune (numerico)             852
Comune                               852
Comune (dizione straniera)        564840
Sigla automobilistica               7473
Capoluogo di Pro

In [22]:
#Creo il df con solo le colonne di interesse

df_final_clean= df_final_raw[["REF_AREA","DATA_TYPE","RESULT","TIME_PERIOD","OBS_VALUE","Codice Regione", "Codice Provincia/Uts",
                      "Codice Comune (numerico)",
                      "Comune","Capoluogo di Provincia/Uts", "Capoluogo di Regione", "Superficie (Kmq)", 
                      "Popolazione residente", "Anno (Popolazione residente)", "_merge"]]
df_final_clean

,REF_AREA,DATA_TYPE,RESULT,TIME_PERIOD,OBS_VALUE,Codice Regione,Codice Provincia/Uts,Codice Comune (numerico),Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente,Anno (Popolazione residente),_merge
0,1001,KILLINJ,F,2001,10,1.0,1.0,1001.0,Agliè,0.0,0.0,"13,1462",2557.0,2001.0,both
1,1001,KILLINJ,F,2002,10,1.0,1.0,1001.0,Agliè,0.0,0.0,"13,1462",2538.0,2002.0,both
2,1001,KILLINJ,F,2003,7,1.0,1.0,1001.0,Agliè,0.0,0.0,"13,1462",2588.0,2003.0,both
3,1001,KILLINJ,F,2004,13,1.0,1.0,1001.0,Agliè,0.0,0.0,"13,1462",2679.0,2004.0,both
4,1001,KILLINJ,F,2005,2,1.0,1.0,1001.0,Agliè,0.0,0.0,"13,1462",2674.0,2005.0,both
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
573547,111107,ROADACC,9,2020,2,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,"27,1937",2549.0,2020.0,both
573548,111107,ROADACC,9,2021,5,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,"27,1943",2536.0,2021.0,both
573549,111107,ROADACC,9,2022,1,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,"27,1943",2575.0,2022.0,both
573550,111107,ROADACC,9,2023,4,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,"27,1943",2616.0,2023.0,both


In [23]:
#voglio vedere quanti comuni univoci c' erano nel 2020
df_final_clean[df_final_clean["TIME_PERIOD"] == 2020]["REF_AREA"].nunique()

7904

In [24]:
#PRIMA DI ELIMINARE LA COLONNA MERGE MI RICREO IL DF CON SOLO QUELLI CHE SI SONO MATCHATI
#SI VANNO A PERDERE SOLO 850 RECORD CIRCA SU PIù DI 500000

df_final_clean=df_final_clean[df_final_clean["_merge"] == "both"].copy()

In [25]:
#Elimino merge che non mi serve più, il codice d
df_final_clean=df_final_clean.drop(columns=["_merge","Anno (Popolazione residente)","Codice Comune (numerico)"]) #ho dovuto riscrivere df_final_clean= perchè drop non eliminerebbe 
                                                            #in maniera permanente la colonna, se non volessi salvare il 
                                                            #df come un nuovo df, dovrei mettere drop(columns=["_merge"], inplace=True)

df_final_clean

,REF_AREA,DATA_TYPE,RESULT,TIME_PERIOD,OBS_VALUE,Codice Regione,Codice Provincia/Uts,Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente
0,1001,KILLINJ,F,2001,10,1.0,1.0,Agliè,0.0,0.0,"13,1462",2557.0
1,1001,KILLINJ,F,2002,10,1.0,1.0,Agliè,0.0,0.0,"13,1462",2538.0
2,1001,KILLINJ,F,2003,7,1.0,1.0,Agliè,0.0,0.0,"13,1462",2588.0
3,1001,KILLINJ,F,2004,13,1.0,1.0,Agliè,0.0,0.0,"13,1462",2679.0
4,1001,KILLINJ,F,2005,2,1.0,1.0,Agliè,0.0,0.0,"13,1462",2674.0
...,...,...,...,...,...,...,...,...,...,...,...,...
573547,111107,ROADACC,9,2020,2,20.0,111.0,Villaspeciosa,0.0,0.0,"27,1937",2549.0
573548,111107,ROADACC,9,2021,5,20.0,111.0,Villaspeciosa,0.0,0.0,"27,1943",2536.0
573549,111107,ROADACC,9,2022,1,20.0,111.0,Villaspeciosa,0.0,0.0,"27,1943",2575.0
573550,111107,ROADACC,9,2023,4,20.0,111.0,Villaspeciosa,0.0,0.0,"27,1943",2616.0


In [26]:
df_final_clean.isna().sum()

REF_AREA                      0
DATA_TYPE                     0
RESULT                        0
TIME_PERIOD                   0
OBS_VALUE                     0
Codice Regione                0
Codice Provincia/Uts          0
Comune                        0
Capoluogo di Provincia/Uts    0
Capoluogo di Regione          0
Superficie (Kmq)              0
Popolazione residente         0
dtype: int64

In [27]:
#controllo i tipi di dato
df_final_clean.info()

<class 'pandas.DataFrame'>
Index: 572700 entries, 0 to 573551
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   REF_AREA                    572700 non-null  int64  
 1   DATA_TYPE                   572700 non-null  str    
 2   RESULT                      572700 non-null  str    
 3   TIME_PERIOD                 572700 non-null  int64  
 4   OBS_VALUE                   572700 non-null  int64  
 5   Codice Regione              572700 non-null  float64
 6   Codice Provincia/Uts        572700 non-null  float64
 7   Comune                      572700 non-null  str    
 8   Capoluogo di Provincia/Uts  572700 non-null  float64
 9   Capoluogo di Regione        572700 non-null  float64
 10  Superficie (Kmq)            572700 non-null  str    
 11  Popolazione residente       572700 non-null  float64
dtypes: float64(5), int64(3), str(4)
memory usage: 56.8 MB


In [28]:
#SISTEMO I TIPI DELLE COLONNE

df_final_clean["Codice Regione"] = (df_final_clean["Codice Regione"].astype("Int64"))
# Int64 (con la I maiuscola) permette di avere interi e valori mancanti nella stessa colonna.
df_final_clean["Codice Provincia/Uts"] = (df_final_clean["Codice Provincia/Uts"].astype("Int64"))
df_final_clean["Capoluogo di Provincia/Uts"] = (df_final_clean["Capoluogo di Provincia/Uts"].astype("Int64"))
df_final_clean["Capoluogo di Regione"] = (df_final_clean["Capoluogo di Regione"].astype("Int64"))
df_final_clean["Popolazione residente"] = (df_final_clean["Popolazione residente"].astype("Int64"))

In [29]:
df_final_clean[df_final_clean["Superficie (Kmq)"].str.contains("\.", regex=True)] #così mi dà come risultato tutto il df con la colonna sup con i caratteri contenenti .

df_final_clean[df_final_clean["Superficie (Kmq)"].str.contains("\.", regex=True)]["Superficie (Kmq)"] #così mi fa vedere solo la colonna Superficie con i caratteri contenti .
#N.B. ricorda lo \ prima del . altrimenti cercherebbe una concatenazione di parole e non proprio il punto come valore

#regex=False -->Interpreta la stringa come testo letterale, quindi i caratteri speciali sono esattamente quello che sono 
                #a livello letterale. Serve per avere una stringa esatta
#regex=True --> quando vuoi sfruttare i simboli delle espressioni regolari (\s (spazio), \d, ^, $, ., +, *, ecc.) e 
                # non che sia letterale il carattere speciale

348585    1.287,3372
348586    1.287,3371
348587    1.287,3371
348588    1.287,3371
348589    1.287,3371
             ...    
348652    1.287,3716
348653    1.288,1875
348654    1.288,1875
348655    1.288,1875
348656    1.288,1875
Name: Superficie (Kmq), Length: 72, dtype: str

In [30]:
df_final_clean["Superficie (Kmq)"] = (df_final_clean["Superficie (Kmq)"].str.replace(".", "", regex=False)
                                        .str.replace(",", ".", regex=False)
                                        .astype(float))

In [31]:
df_final_clean["Superficie (Kmq)"]

0         13.1462
1         13.1462
2         13.1462
3         13.1462
4         13.1462
           ...   
573547    27.1937
573548    27.1943
573549    27.1943
573550    27.1943
573551    27.1943
Name: Superficie (Kmq), Length: 572700, dtype: float64

In [32]:
print(df_final_clean.shape) #righe, colonne
print("-"*50)
print(df_final_clean.info())

(572700, 12)
--------------------------------------------------
<class 'pandas.DataFrame'>
Index: 572700 entries, 0 to 573551
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   REF_AREA                    572700 non-null  int64  
 1   DATA_TYPE                   572700 non-null  str    
 2   RESULT                      572700 non-null  str    
 3   TIME_PERIOD                 572700 non-null  int64  
 4   OBS_VALUE                   572700 non-null  int64  
 5   Codice Regione              572700 non-null  Int64  
 6   Codice Provincia/Uts        572700 non-null  Int64  
 7   Comune                      572700 non-null  str    
 8   Capoluogo di Provincia/Uts  572700 non-null  Int64  
 9   Capoluogo di Regione        572700 non-null  Int64  
 10  Superficie (Kmq)            572700 non-null  float64
 11  Popolazione residente       572700 non-null  Int64  
dtypes: Int64(5), float64(1),

In [33]:
df_final_clean

,REF_AREA,DATA_TYPE,RESULT,TIME_PERIOD,OBS_VALUE,Codice Regione,Codice Provincia/Uts,Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente
0,1001,KILLINJ,F,2001,10,1,1,Agliè,0,0,13.1462,2557
1,1001,KILLINJ,F,2002,10,1,1,Agliè,0,0,13.1462,2538
2,1001,KILLINJ,F,2003,7,1,1,Agliè,0,0,13.1462,2588
3,1001,KILLINJ,F,2004,13,1,1,Agliè,0,0,13.1462,2679
4,1001,KILLINJ,F,2005,2,1,1,Agliè,0,0,13.1462,2674
...,...,...,...,...,...,...,...,...,...,...,...,...
573547,111107,ROADACC,9,2020,2,20,111,Villaspeciosa,0,0,27.1937,2549
573548,111107,ROADACC,9,2021,5,20,111,Villaspeciosa,0,0,27.1943,2536
573549,111107,ROADACC,9,2022,1,20,111,Villaspeciosa,0,0,27.1943,2575
573550,111107,ROADACC,9,2023,4,20,111,Villaspeciosa,0,0,27.1943,2616


In [34]:
#Rinomino i valori per renderli più chiari
df_final_clean["DATA_TYPE"] = df_final_clean["DATA_TYPE"].replace({"KILLINJ": "morti_e_feriti",
                                                                    "ROADACC": "incidenti_stradali"})

df_final_clean["RESULT"] = df_final_clean["RESULT"].replace({"M": "morti",
                                                            "F": "feriti",
                                                            "9": "incidenti"})

df_final_clean=df_final_clean.rename(columns={"REF_AREA" : "ID_COMUNE", 
                                                              "Codice Provincia/Uts" : "ID_PROVINCIA",
                                                              "Codice Regione" : "ID_REGIONE",
                                                              "Capoluogo di Provincia/Uts": "CAPOLUOGO PROVINCIA",
                                                              "Capoluogo di Regione" : "CAPOLUOGO REGIONE",
                                                              "Comune" : "COMUNE",
                                                              "Popolazione residente" : "RESIDENTI",
                                                              "Superficie (Kmq)" : "SUPERFICIE (KMQ)",})
df_final_clean

,ID_COMUNE,DATA_TYPE,RESULT,TIME_PERIOD,OBS_VALUE,ID_REGIONE,ID_PROVINCIA,COMUNE,CAPOLUOGO PROVINCIA,CAPOLUOGO REGIONE,SUPERFICIE (KMQ),RESIDENTI
0,1001,morti_e_feriti,feriti,2001,10,1,1,Agliè,0,0,13.1462,2557
1,1001,morti_e_feriti,feriti,2002,10,1,1,Agliè,0,0,13.1462,2538
2,1001,morti_e_feriti,feriti,2003,7,1,1,Agliè,0,0,13.1462,2588
3,1001,morti_e_feriti,feriti,2004,13,1,1,Agliè,0,0,13.1462,2679
4,1001,morti_e_feriti,feriti,2005,2,1,1,Agliè,0,0,13.1462,2674
...,...,...,...,...,...,...,...,...,...,...,...,...
573547,111107,incidenti_stradali,incidenti,2020,2,20,111,Villaspeciosa,0,0,27.1937,2549
573548,111107,incidenti_stradali,incidenti,2021,5,20,111,Villaspeciosa,0,0,27.1943,2536
573549,111107,incidenti_stradali,incidenti,2022,1,20,111,Villaspeciosa,0,0,27.1943,2575
573550,111107,incidenti_stradali,incidenti,2023,4,20,111,Villaspeciosa,0,0,27.1943,2616


In [ ]:
#AGGIUNGO IL NOME DELLE REGIONI

regioni = {
    1: "Piemonte",
    2: "Valle d'Aosta",
    3: "Lombardia",
    4: "Trentino-Alto Adige",
    5: "Veneto",
    6: "Friuli-Venezia Giulia",
    7: "Liguria",
    8: "Emilia-Romagna",
    9: "Toscana",
    10: "Umbria",
    11: "Marche",
    12: "Lazio",
    13: "Abruzzo",
    14: "Molise",
    15: "Campania",
    16: "Puglia",
    17: "Basilicata",
    18: "Calabria",
    19: "Sicilia",
    20: "Sardegna"}

df_final_clean["REGIONE"] = df_final_clean["ID_REGIONE"].map(regioni)

In [36]:
df_final_clean["RESULT"].value_counts()
#Ho provato a ricontrollare anche in df_istat_raw e il rapporto è sempre questo, 1:1:1

RESULT
feriti       190900
morti        190900
incidenti    190900
Name: count, dtype: int64

In [37]:
os.getcwd()
os.listdir()

['00_FIXED_fetching.ipynb',
 '01_FIXED_cleanind_overview.ipynb',
 '02a.EDA_&_Temporal_analysis.ipynb',
 '02b_EDA_Municipality_analysis_aggregate_01_24.ipynb',
 '03b.Test.ipynb',
 'APPUNTI.txt',
 'Df_clean',
 'Df_raw',
 'FIXED_fetching data',
 'FUNCTIONS.py',
 '__pycache__']

In [38]:
if not os.path.exists("Df_clean"):
    os.mkdir("Df_clean")
df_final_clean.to_csv("Df_clean/df_final_from_01_to_24_clean.csv",index=False)

In [39]:
df_incidenti= df_final_clean[df_final_clean["RESULT"]=="incidenti"]
df_incidenti.to_csv("Df_clean/df_incidenti_from_01_to_24.csv",index=False)

df_morti= df_final_clean[df_final_clean["RESULT"]=="morti"]
df_morti.to_csv("Df_clean/df_morti_from_01_to_24.csv",index=False)

df_feriti= df_final_clean[df_final_clean["RESULT"]=="feriti"]
df_feriti.to_csv("Df_clean/df_feriti_from_01_to_24.csv",index=False)